tokio/src/runtime/io/driver/uring.rs

```diff
@@ -123,23 +182,70 @@ impl Drop for UringContext {
 
 impl Handle {
     #[allow(dead_code)]
-    pub(crate) fn add_uring_source(&self, interest: Interest) -> io::Result<()> {
-        // setup for io_uring
-        let uringfd = self.get_uring().lock().uring.as_raw_fd();
+    fn add_uring_source(&self, uringfd: RawFd) -> io::Result<()> {
         let mut source = SourceFd(&uringfd);
         self.registry
-            .register(&mut source, TOKEN_WAKEUP, interest.to_mio())
+            .register(&mut source, TOKEN_WAKEUP, Interest::READABLE.to_mio())
     }
 
     pub(crate) fn get_uring(&self) -> &Mutex<UringContext> {
         &self.uring_context
     }
 
+    fn set_uring_state(&self, state: State) {
+        self.uring_state.store(state.as_usize(), Ordering::Release);
+    }
+
+    /// Check if the io_uring context is initialized. If not, it will try to initialize it.
+    pub(crate) fn check_and_init(&self) -> io::Result<bool> {
+        match State::from_usize(self.uring_state.load(Ordering::Acquire)) {
```

Comment:
nit: Just curious, what is the reason for caching `uring_state` at the runtime-level?

Reply:
`State` keeps track of whether the uring driver has been initialized (`io_uring_setup`), and since the uring driver is tied to the runtime, cache is currently maintained for each runtime.

Reply:
Not a blocker.

Are there difficulties or overhead to initialize the io_uring while building the runtime?

It looks like the `check_and_init` will be called (by `register_op()` frequently, I'm concern about the performance overhead of atomic load in hot path. Can this a be thread-local value?

Reply:
> Are there difficulties or overhead to initialize the io_uring while building the runtime?

No, there aren't any major difficulties. I think lazy initialization of uring is not strictly required here, but it has the advantage of not having to perform extra syscalls when io_uring is not actually performed.


> It looks like the `check_and_init` will be called (by `register_op()` frequently, I'm concern about the performance overhead for loading atomic value in hot path.

As long as we check if io_uring is supported at *runtime*, I think we can't remove this check. Also, I don't expect loading an atomic variable to have a huge performance impact.

We could also enforce io_uring usage at compile time and remove runtime check, but that has trade-offs too.